codec-dns/src/main/java/io/netty/handler/codec/dns/DatagramDnsQueryDecoder.java

```diff
@@ -0,0 +1,111 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.dns;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.socket.DatagramPacket;
+import io.netty.handler.codec.CorruptedFrameException;
+import io.netty.handler.codec.MessageToMessageDecoder;
+
+import java.net.InetSocketAddress;
+import java.util.List;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+
+/**
+ * Decodes a {@link DatagramPacket} into a {@link DatagramDnsQuery}.
+ */
+@ChannelHandler.Sharable
+public class DatagramDnsQueryDecoder extends MessageToMessageDecoder<DatagramPacket> {
+
+    private final DnsRecordDecoder recordDecoder;
+
+    /**
+     * Creates a new decoder with {@linkplain DnsRecordDecoder#DEFAULT the default record decoder}.
+     */
+    public DatagramDnsQueryDecoder() {
+        this(DnsRecordDecoder.DEFAULT);
+    }
+
+    /**
+     * Creates a new decoder with the specified {@code recordDecoder}.
+     */
+    public DatagramDnsQueryDecoder(DnsRecordDecoder recordDecoder) {
+        this.recordDecoder = checkNotNull(recordDecoder, "recordDecoder");
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, DatagramPacket packet, List<Object> out) throws Exception {
+        final InetSocketAddress sender = packet.sender();
+        final ByteBuf buf = packet.content();
+
+        final DnsQuery query = newQuery(sender, buf);
+        boolean success = false;
+        try {
+            final int questionCount = buf.readUnsignedShort();
+            final int answerCount = buf.readUnsignedShort();
+            final int authorityRecordCount = buf.readUnsignedShort();
+            final int additionalRecordCount = buf.readUnsignedShort();
+
+            decodeQuestions(query, buf, questionCount);
+            decodeRecords(query, DnsSection.ANSWER, buf, answerCount);
+            decodeRecords(query, DnsSection.AUTHORITY, buf, authorityRecordCount);
+            decodeRecords(query, DnsSection.ADDITIONAL, buf, additionalRecordCount);
+
+            out.add(query);
+            success = true;
+        } finally {
+            if (!success) {
+                query.release();
+            }
+        }
+    }
+
+    private static DnsQuery newQuery(InetSocketAddress sender, ByteBuf buf) {
+        final int id = buf.readUnsignedShort();
+
+        final int flags = buf.readUnsignedShort();
+        if (flags >> 15 == 1) {
+            throw new CorruptedFrameException("not a query");
+        }
+        final DnsQuery query =
+            new DatagramDnsQuery(sender, null, id, DnsOpCode.valueOf((byte) (flags >> 11 & 0xf)));
```

Comment:
Could you pass recipient here using `packet.recipient()`? It lets a user know from which bind address (or interface) the server received the query from.